modules/core/src/lib/effect.ts

```diff
@@ -8,17 +9,32 @@ export type PostRenderOptions = LayersPassRenderOptions & {
   inputBuffer: Framebuffer;
   swapBuffer: Framebuffer;
 };
+export type EffectContext = {
```

Comment:
`device` can be obtained from `deck`. What is the thinking to include them separately? Easier testing?

I see in the followup that `deck` is needed to [add/remove shader modules](https://github.com/visgl/deck.gl/pull/8682/files#diff-d41f5114076a8be933e2409af8aa332e45ecb5e146ac71c18375229995f8ba31R678-R685). If this is the sole purpose of including `deck`, perhaps it would be better to pass `layerManager` directly:

```ts
export type EffectContext = {
  device: Device;
  layerManager: LayerManager;
};
```

Reply:
- `Deck.device` is not public.
- It's intentional to route everything through deck, so *manager do not interact with each other directly. If we do change the implementation such as who tracks the shader modules, there will be no breaking change to the effects.